### PR TITLE
Workaround for complex Expression as dict key issue

### PR DESCRIPTION
**Pull request by demianw**
There is a code section in https://github.com/NeuroLang/NeuroLang/blob/1e55827573037bea541240654c6ae9e2bbb68e56/neurolang/probabilistic/dalvi_suciu_lift.py#L549
where there is a bug in which complex logic expressions don't work well as keys of a dictionary providing random results in the equality which we can't reproduce outside the dictionary methods (we don't get why).  This is shown in #652. This PR aims at correcting that by using indexed lists instead of a dictionary and to update the equality of `Symbol` and `Constant` instances as that would have been a part of the issue.

### CODE REVIEW

**Comment by demianw** [comment]
@tgy do you have a test handy to see if this is working now?

**Comment by demianw** [comment]
@gzanitti can you review this PR please? Thanks

**Comment by gzanitti** [comment]
> @gzanitti can you review this PR please? Thanks

sure, but I don't know exactly what I'm reviewing. No description, nothing.
Can I have some context?

**Comment by demianw** [comment]
> > @gzanitti can you review this PR please? Thanks
> 
> sure, but I don't know exactly what I'm reviewing. No description, nothing.
> Can I have some context?

@gzanitti updated the description. Thanks!

File: neurolang/probabilistic/dalvi_suciu_lift.py [hunk 1]
```diff
@@ -584,22 +587,17 @@ def mobius_function(formula, formula_containments, known_weights=None):
         known_weights = dict()
```
Comment (gzanitti) [inline neurolang/probabilistic/dalvi_suciu_lift.py:587]: Only minor comment, is that, looking at the code I don't see any use case for `known_weights` to be an optional parameter, but maybe it is useful outside the scope of these two functions (I mention it for double checking).

Everything else looks good to me.

**Review by gzanitti — APPROVED** [review]

**Pull request merged**